File: jar/src/main/java/jline/api/cache/Cache_ttl_lrua.java
/**
 * @file TTL approximation for LRU caches with arrival routing (access graphs)
 *
 * Characteristic-time (TTL) approximation for LRU caches whose lists form an
 * arbitrary access graph (linear chains and trees), mirroring the MATLAB
 * reference cache_ttl_lrua.m. Each item is modeled by an embedded DTMC over
 * the lists, with transition probabilities given by exponential timer races,
 * converted to time-stationary probabilities via mean holding times. The
 * characteristic times are fixed so that the expected occupancy of each list
 * equals its capacity; the joint fixed point is computed by a damped Newton
 * iteration on the log characteristic times with a finite-difference
 * Jacobian (no external nonlinear solver required).
 *
 * @since LINE 3.0
 */
package jline.api.cache;

import org.apache.commons.math3.util.FastMath;

import jline.api.mc.Dtmc_solve;
import jline.util.matrix.Matrix;

public final class Cache_ttl_lrua {
    private Cache_ttl_lrua() {}

    private static final int MAX_NEWTON = 200;
    private static final double TOL = 1e-10;

    /**
     * Solve LRU caches with access graphs using the TTL tree approximation.
     *
     * @param lambda Per-user arrival rate matrices, lambda[v] is (n x h+1)
     *               with column 0 the out-of-cache request rate and column l
     *               the request rate seen while the item is in list l.
     * @param R      Access graph, R[v][k] is the (h+1 x h+1) routing matrix
     *               of item k for user v (row/col 0 is the out-of-cache node).
     * @param m      Cache capacity vector (1 x h).
     * @return Matrix (n x h+1) of per-item probabilities; column 0 is the
     *         out-of-cache (miss) probability, column l the probability of
     *         residing in list l.
     */
    public static Matrix cache_ttl_lrua(Matrix[] lambda, Matrix[][] R, Matrix m) {
        if (lambda == null || R == null || m == null) {
            throw new IllegalArgumentException("Lambda, R, and m parameters cannot be null");
        }

        int n = lambda[0].getNumRows();
        int h = lambda[0].getNumCols() - 1;

        // Damped Newton on y = log(T): solve occ(exp(y)) - m = 0 jointly.
        // The log parameterization keeps the characteristic times positive.
        double[] y = new double[h];
        double[] res = residual(lambda, R, y, m, n, h);
        double nrm = norm(res);
        for (int it = 0; it < MAX_NEWTON && nrm > TOL; it++) {
            // finite-difference Jacobian
            double[][] J = new double[h][h];
            double eps = 1e-6;
            for (int c = 0; c < h; c++) {
                double[] yp = new double[h];
                System.arraycopy(y, 0, yp, 0, h);
                yp[c] += eps;
                double[] rp = residual(lambda, R, yp, m, n, h);
                for (int r = 0; r < h; r++) {
                    J[r][c] = (rp[r] - res[r]) / eps;
                }
            }
            double[] step = solveLinear(J, res, h);
            if (step == null) {
                break;
            }
            // backtracking line search on the residual norm
            double alpha = 1.0;
            double[] ynew = new double[h];
            double[] rnew = null;
            double nrmNew = Double.POSITIVE_INFINITY;
            for (int ls = 0; ls < 40; ls++) {
                for (int l = 0; l < h; l++) {
                    ynew[l] = y[l] - alpha * step[l];
                }
                rnew = residual(lambda, R, ynew, m, n, h);
                nrmNew = norm(rnew);
                if (nrmNew < nrm) {
                    break;
                }
                alpha *= 0.5;
            }
            if (nrmNew >= nrm) {
                break;
            }
            System.arraycopy(ynew, 0, y, 0, h);
            res = rnew;
            nrm = nrmNew;
        }

        double[] x = new double[h];
        for (int l = 0; l < h; l++) {
            x[l] = FastMath.exp(y[l]);
        }
        return randProb(lambda, R, x, n, h);
    }

    /* Occupancy residuals occ_l(exp(y)) - m_l. */
    private static double[] residual(Matrix[] lambda, Matrix[][] R, double[] y,
                                     Matrix m, int n, int h) {
        double[] x = new double[h];
        for (int l = 0; l < h; l++) {
            x[l] = FastMath.exp(y[l]);
        }
        Matrix prob = randProb(lambda, R, x, n, h);
        double[] res = new double[h];
        for (int l = 0; l < h; l++) {
            double capa = 0.0;
            for (int i = 0; i < n; i++) {
                capa += prob.get(i, l + 1);
            }
            res[l] = capa - m.get(l);
        }
        return res;
    }

    private static double norm(double[] v) {
        double s = 0.0;
        for (int i = 0; i < v.length; i++) {
            s += v[i] * v[i];
        }
        return FastMath.sqrt(s);
    }

    /* Solve J s = r by Gaussian elimination with partial pivoting. */
    private static double[] solveLinear(double[][] J, double[] r, int h) {
        double[][] a = new double[h][h + 1];
        for (int i = 0; i < h; i++) {
            System.arraycopy(J[i], 0, a[i], 0, h);
            a[i][h] = r[i];
        }
        for (int c = 0; c < h; c++) {
            int piv = c;
            for (int i = c + 1; i < h; i++) {
                if (FastMath.abs(a[i][c]) > FastMath.abs(a[piv][c])) {
                    piv = i;
                }
            }
            if (FastMath.abs(a[piv][c]) < 1e-300) {
                return null;
            }
            double[] tmp = a[c];
            a[c] = a[piv];
            a[piv] = tmp;
            for (int i = c + 1; i < h; i++) {
                double f = a[i][c] / a[c][c];
                for (int j = c; j <= h; j++) {
                    a[i][j] -= f * a[c][j];
                }
            }
        }
        double[] s = new double[h];
        for (int i = h - 1; i >= 0; i--) {
            double v = a[i][h];
            for (int j = i + 1; j < h; j++) {
                v -= a[i][j] * s[j];
            }
            s[i] = v / a[i][i];
        }
        return s;
    }

    // see _kb/03-api-layer.md for rationale
    private static Matrix randProb(Matrix[] lambda, Matrix[][] R, double[] x, int n, int h) {
        Matrix randprob = new Matrix(n, h + 1);
        for (int i = 0; i < n; i++) {
            Matrix Ri = R[0][i];
            double[][] trans = new double[h + 1][h + 1];
            for (int j = 0; j <= h; j++) {
                for (int k = 0; k <= h; k++) {
                    if (Ri.get(j, k) > 0) {
                        if (j == 0) {
                            trans[j][k] = Ri.get(j, k);
                        } else {
                            trans[j][k] = (1.0 - FastMath.exp(-lambda[0].get(i, j) * x[j - 1]))
                                    * Ri.get(j, k);
                        }
                        if (j != k && k > 0) {
                            trans[k][j] = FastMath.exp(-lambda[0].get(i, k) * x[k - 1]);
                        }
                    }
                }
            }

            // see _kb/03-api-layer.md for rationale
            boolean[] conn = new boolean[h + 1];
            int nconn = 0;
            for (int j = 0; j <= h; j++) {
                for (int k = 0; k <= h; k++) {
                    if (trans[j][k] != 0.0) {
                        conn[k] = true;
                    }
                }
            }
            int[] map = new int[h + 1];
            for (int j = 0; j <= h; j++) {
                if (conn[j]) {
                    map[j] = nconn;
                    nconn++;
                } else {
                    map[j] = -1;
                }
            }
            if (nconn == 0) {
                continue;
            }
            Matrix sub = new Matrix(nconn, nconn);
            for (int j = 0; j <= h; j++) {
                if (!conn[j]) {
                    continue;
                }
                for (int k = 0; k <= h; k++) {
                    if (conn[k]) {
                        sub.set(map[j], map[k], trans[j][k]);
                    }
                }
            }
            Matrix dtmcprob = Dtmc_solve.dtmc_solve(sub);

            double denom = 0.0;
            double[] avgtime = new double[h + 1];
            double[] ssprob = new double[h + 1];
            for (int node = 0; node <= h; node++) {
                if (!conn[node]) {
                    continue;
                }
                ssprob[node] = dtmcprob.get(map[node]);
                double rate = lambda[0].get(i, node);
                if (node > 0) {
                    avgtime[node] = rate > 0
                            ? (1.0 - FastMath.exp(-rate * x[node - 1])) / rate : 0.0;
                } else {
                    avgtime[node] = rate > 0 ? 1.0 / rate : 1.0;
                }
                denom += ssprob[node] * avgtime[node];
            }
            if (denom <= 0) {
                continue;
            }
            for (int node = 0; node <= h; node++) {
                if (conn[node]) {
                    randprob.set(i, node, ssprob[node] * avgtime[node] / denom);
                }
            }
        }
        return randprob;
    }
}

File: jar/src/main/kotlin/jline/api/cache/Cache_ttl_lrua.kt
/**
 * @file TTL-based LRU(A) Cache Analysis
 * 
 * Implementation of Time-To-Live (TTL) approximation for LRU(A) cache systems 
 * using tree-structured replacement policies. This method enables efficient 
 * analysis of multi-level cache hierarchies with arbitrary routing patterns.
 * 
 * @since LINE 3.0
 */
package jline.api.cache

import jline.util.matrix.Matrix
import org.apache.commons.math3.optim.PointValuePair
import org.apache.commons.math3.optim.nonlinear.scalar.GoalType
import org.apache.commons.math3.optim.nonlinear.scalar.noderiv.NelderMeadSimplex
import org.apache.commons.math3.optim.nonlinear.scalar.noderiv.SimplexOptimizer
import org.apache.commons.math3.optim.nonlinear.scalar.ObjectiveFunction
import org.apache.commons.math3.optim.InitialGuess
import org.apache.commons.math3.optim.MaxEval
import org.apache.commons.math3.analysis.MultivariateFunction
import org.apache.commons.math3.util.FastMath
import jline.api.mc.dtmc_solve
import java.util.Random

/**
 * Solve arbitrary replacement policy caches LRU(A) using the TTL tree approximation.
 *
 * @param lambda - Array of matrices representing arrival rates. Dimensions: [u][n][h+1] where
 *                 u = number of users, n = number of items, h = number of lists
 * @param R - Array of routing matrices. Dimensions: [u][n] where each R[u][n] is a (h+1)x(h+1) matrix
 *            representing the tree structure for item n from user u
 * @param m - Matrix representing cache capacity vector for each list
 * @return Matrix - Steady-state probabilities for each item at each list
 */
fun cache_ttl_lrua(lambda: Array<Matrix?>?, R: Array<Array<Matrix?>?>?, m: Matrix?): Matrix {
    if (lambda == null || R == null || m == null) {
        throw IllegalArgumentException("Lambda, R, and m parameters cannot be null")
    }
    
    val u = lambda.size // number of users (should be 1 for single-user case)
    val n = lambda[0]!!.numCols // number of items
    val h = lambda[0]!!.numRows - 1 // number of lists
    
    // Initialize random number generator
    val rand = Random()
    
    // Generate initial values for list times
    val rangeLeft = 0.0
    val rangeRight = 10.0
    val initialX = DoubleArray(h) { 
        rand.nextDouble() * (rangeRight - rangeLeft) + rangeLeft
    }
    
    // Define the objective function for optimization
    val objectiveFunction = MultivariateFunction { x ->
        val (F, _, _) = ttlTreeTime(x, lambda, R, m, n, h)
        // Return the norm of F as the objective to minimize
        var sum = 0.0
        for (i in 0 until F.numCols) {
            sum += F[0, i] * F[0, i]
        }
        FastMath.sqrt(sum)
    }
    
    // Set up optimizer
    val optimizer = SimplexOptimizer(1e-10, 1e-30)
    val simplex = NelderMeadSimplex(h)
    
    // Optimize to find list times
    val optimum: PointValuePair = optimizer.optimize(
        MaxEval(1000000),
        ObjectiveFunction(objectiveFunction),
        GoalType.MINIMIZE,
        InitialGuess(initialX),
        simplex
    )
    
    val listTime = optimum.point
    
    // Get steady-state probabilities using the optimized list times
    val (_, ssProb, _) = ttlTreeTime(listTime, lambda, R, m, n, h)
    
    return ssProb
}

/**
 * Helper function to compute TTL tree time equations
 */
private fun ttlTreeTime(
    x: DoubleArray, 
    lambda: Array<Matrix?>?, 
    R: Array<Array<Matrix?>?>?, 
    m: Matrix?,
    n: Int, 
    h: Int
): Triple<Matrix, Matrix, Matrix> {
    val steadyStateProb = Matrix(n, h + 1)
    val randProb = Matrix(n, h + 1)
    val avgTime = Matrix(n, h + 1)
    val cDiff = Matrix(1, h)
    val capa = Matrix(1, h)
    val rpDenominator = Matrix(1, n)
    
    // Calculate probability of each item at each list
    for (i in 0 until n) {
        val transMatrix = Matrix(h + 1, h + 1)
        
        // Build transition matrix for item i
        for (j in 0..h) {
            // Find leaf nodes (non-zero entries in row j of routing matrix)
            val leafNodes = mutableListOf<Int>()
            for (k in 0..h) {
                if (R!![0]!![i]!![j, k] > 0) {
                    leafNodes.add(k)
                }
            }
            
            for (k in leafNodes) {
                if (j == 0) {
                    transMatrix[j, k] = R!![0]!![i]!![j, k]
                } else {
                    transMatrix[j, k] = (1 - FastMath.exp(-lambda!![0]!![j, i] * x[j - 1])) * R!![0]!![i]!![j, k]
                }
                if (j != k) {
                    transMatrix[k, j] = FastMath.exp(-lambda!![0]!![k, i] * x[k - 1])
                }
            }
        }
        
        // Find connected components and remove disconnected nodes
        val missConnection = mutableListOf<Int>()
        for (row in 0..h) {
            var allZero = true
            for (col in 0..h) {
                if (transMatrix[row, col] != 0.0) {
                    allZero = false
                    break
                }
            }
            if (allZero) {
                missConnection.add(row)
            }
        }
        
        val dtChain = (0..h).filter { it !in missConnection }
        
        if (dtChain.isNotEmpty()) {
            // Create reduced transition matrix
            val reducedMatrix = Matrix(dtChain.size, dtChain.size)
            for ((idxA, a) in dtChain.withIndex()) {
                for ((idxB, b) in dtChain.withIndex()) {
                    reducedMatrix[idxA, idxB] = transMatrix[a, b]
                }
            }
            
            // Solve DTMC for steady-state probabilities
            val dtmcProb = dtmc_solve(reducedMatrix)
            
            // Assign probabilities and calculate average times
            for ((idx, a) in dtChain.withIndex()) {
                steadyStateProb[i, a] = dtmcProb[0, idx]
                
                if (a > 0) {
                    avgTime[i, a] = (1 - FastMath.exp(-lambda!![0]!![a, i] * x[a - 1])) / lambda!![0]!![a, i]
                } else {
                    avgTime[i, a] = 1.0 / lambda!![0]!![a, i]
                }
                
                rpDenominator[0, i] += steadyStateProb[i, a] * avgTime[i, a]
            }
            
            // Calculate random time probabilities
            for (a in dtChain) {
                randProb[i, a] = steadyStateProb[i, a] * avgTime[i, a] / rpDenominator[0, i]
            }
        }
    }
    
    // Calculate capacity differences
    for (l in 0 until h) {
        for (i in 0 until n) {
            capa[0, l] += randProb[i, l + 1]
        }
        cDiff[0, l] = m!![0, l] - capa[0, l]
    }
    
    return Triple(cDiff, randProb, cDiff)
}
/**
 * Cache ttl lrua algorithms
 */
@Suppress("unused")
class CacheTtlLruaAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}